cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1690,175 +1853,207 @@ private predicate flowFwdStore(
     ap instanceof AccessPathNil and
     flowFwd(node, _, _, ap, config)
   )
   or
-  exists(Content f, AccessPath ap0 |
-    flowRead(node, f, toReturn, ap0, config) and
-    push(ap0, f, ap)
+  exists(NodeExt mid, AccessPath ap0 |
+    readFwd(node, mid, ap, ap0, config) and
+    flow(mid, toReturn, ap0, config)
   )
   or
-  exists(Node mid, Content f |
+  exists(NodeExt mid, Content f |
     argumentFlowsThrough3(node, mid, _, _, TSummaryReadTaint(f), config) and
     flow(mid, toReturn, any(AccessPathNil nil1), config) and
-    push(any(AccessPathNil nil2), f, ap) and
+    ap = push(f, any(AccessPathNil nil2)) and
     flowFwd(node, _, _, ap, config)
   )
 }
 
 pragma[nomagic]
-private predicate flowStore(
-  Node node, Content f, boolean toReturn, AccessPath ap0, Configuration config
+private predicate storeFwd(
+  NodeExt node1, Content f, NodeExt node2, AccessPath ap, AccessPath ap0, Configuration config
 ) {
-  exists(Node mid |
-    store(node, f, mid) and
-    flow(mid, toReturn, ap0, config)
-  )
+  storeExt(node1, f, node2, _) and
+  flowFwdStore(node2, f, ap, _, _, config) and
+  ap0 = push(f, ap)
 }
 
 pragma[nomagic]
 private predicate flowTaintStore(
-  Node node, Content f, boolean toReturn, AccessPath ap0, Configuration config
+  NodeExt node, Content f, boolean toReturn, AccessPath ap0, Configuration config
 ) {
-  exists(Node mid |
+  exists(NodeExt mid |
     argumentFlowsThrough3(node, mid, _, _, TSummaryTaintStore(f), config) and
     flow(mid, toReturn, ap0, config)
   )
 }
 
 pragma[nomagic]
-private predicate flowRead(
-  Node node, Content f, boolean toReturn, AccessPath ap0, Configuration config
+private predicate readFwd(
+  NodeExt node1, NodeExt node2, AccessPath ap, AccessPath ap0, Configuration config
 ) {
-  exists(Node mid |
-    read(node, f, mid) and
-    flow(mid, toReturn, ap0, config)
+  exists(Content f |
+    readExt(node1, f, node2, _) and
+    flowFwdRead(node2, f, ap, _, config) and
+    ap0 = pop(f, ap)
   )
 }
 
 bindingset[conf, result]
 private Configuration unbind(Configuration conf) { result >= conf and result <= conf }
 
-private predicate flow(Node n, Configuration config) { flow(n, _, _, config) }
+private predicate flow(Node n, Configuration config) { flow(TNormalNode(n), _, _, config) }
 
 private newtype TSummaryCtx =
   TSummaryCtxNone() or
-  TSummaryCtxNil(ParameterNode p) {
-    exists(Configuration conf, ReturnNodeExt ret |
-      flow(p, true, TNil(_), conf) and
-      parameterFlowReturn(p, ret, _, _, _,
-        any(Summary s | s = TSummaryTaint() or s = TSummaryTaintStore(_)), conf) and
-      flow(ret, unbind(conf))
-    )
-  } or
-  TSummaryCtxConsNil(ParameterNode p, Content f) {
-    exists(Configuration conf, ReturnNodeExt ret |
-      flow(p, true, TConsNil(f, _), conf) and
-      parameterFlowReturn(p, ret, _, _, _, TSummaryReadTaint(f), conf) and
-      flow(ret, unbind(conf))
+  TSummaryCtxSome(ParameterNode p, AccessPath ap) {
```

Comment:
We should keep the information in summary context to a bare minimum, since any unnecessary information can cause combinatorial explosions in degenerate cases.  So we should keep the two removed cases and add a third that merely tracks the Content at the head of the access path and e.g. assumes ap length >= 2 to dissallow taint steps.  This will also need some mutation along the path, since we shouldn't be allowed to pop more than one element when collecting such a value-flow-through summary.

Reply:
Access paths have been replaced by untyped access paths in summary contexts.

Reply:
Actually, the more I think about it, I am tempted to go back to regular access paths; both because this makes the code simpler, but also because the extra type information might be used to prune impossible flow paths. @aschackmull : What do you think?